File: examples/cpp/Misc/IrLeds/ir_leds.cpp
#include <algorithm>
#include <atomic>
#include <csignal>
#include <iomanip>
#include <iostream>
#include <memory>
#include <opencv2/opencv.hpp>

#include "depthai/depthai.hpp"

std::atomic<bool> gQuit(false);

void signalHandler(int) {
    gQuit = true;
}

void printDrivers(std::shared_ptr<dai::Device>& device) {
    std::cout << "Detected IR drivers:\n";
    for(const auto& d : device->getIrDrivers()) {
        std::cout << "  name=" << std::get<0>(d)
                  << "  bus=" << std::get<1>(d)
                  << "  addr=0x" << std::hex << std::get<2>(d) << std::dec << "\n";
    }
}

void applyLeds(std::shared_ptr<dai::Device>& device, float dot, float flood, int mask) {
    bool dotOk = device->setIrLaserDotProjectorIntensity(dot, mask);
    bool floodOk = device->setIrFloodLightIntensity(flood, mask);
    std::cout << "dot=" << dot << "  flood=" << flood
              << "  mask=0x" << std::hex << mask << std::dec
              << "  dot_ok=" << dotOk << "  flood_ok=" << floodOk << "\n";
}

int main() {
    signal(SIGINT, signalHandler);
    signal(SIGTERM, signalHandler);

    try {
        auto device = std::make_shared<dai::Device>();
        std::cout << "Connected to: " << device->getDeviceName() << "\n";
        printDrivers(device);

        dai::Pipeline pipeline(device);

        auto monoLeft = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_B);
        auto monoRight = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_C);

        auto leftOut = monoLeft->requestFullResolutionOutput(dai::ImgFrame::Type::NV12);
        auto rightOut = monoRight->requestFullResolutionOutput(dai::ImgFrame::Type::NV12);

        auto leftQueue = leftOut->createOutputQueue();
        auto rightQueue = rightOut->createOutputQueue();

        float dotIntensity = 0.0f;
        float floodIntensity = 0.0f;
        int mask = -1;  // all drivers/sides

        applyLeds(device, dotIntensity, floodIntensity, mask);

        pipeline.start();

        std::cout << "Controls:\n"
                  << "  W/S : increase/decrease dot projector intensity\n"
                  << "  A/D : increase/decrease flood light intensity\n"
                  << "  1   : apply only to LEFT (mask 0x1)\n"
                  << "  2   : apply only to RIGHT (mask 0x2)\n"
                  << "  3   : apply to ALL (mask -1)\n"
                  << "  P   : re-print IR drivers\n"
                  << "  Q   : quit\n";

        while(pipeline.isRunning() && !gQuit) {
            auto left = leftQueue->get<dai::ImgFrame>();
            auto right = rightQueue->get<dai::ImgFrame>();
            if(!left || !right) continue;

            cv::imshow("left", left->getCvFrame());
            cv::imshow("right", right->getCvFrame());

            int key = cv::waitKey(1);
            bool changed = false;

            if(key == 'q' || key == 'Q' || key == 27) {
                break;
            } else if(key == 'w' || key == 'W') {
                dotIntensity = std::min(dotIntensity + 0.1f, 1.0f);
                changed = true;
            } else if(key == 's' || key == 'S') {
                dotIntensity = std::max(dotIntensity - 0.1f, 0.0f);
                changed = true;
            } else if(key == 'a' || key == 'A') {
                floodIntensity = std::min(floodIntensity + 0.1f, 1.0f);
                changed = true;
            } else if(key == 'd' || key == 'D') {
                floodIntensity = std::max(floodIntensity - 0.1f, 0.0f);
                changed = true;
            } else if(key == '1') {
                mask = 0x1;
                changed = true;
            } else if(key == '2') {
                mask = 0x2;
                changed = true;
            } else if(key == '3') {
                mask = -1;
                changed = true;
            } else if(key == 'p' || key == 'P') {
                printDrivers(device);
            }

            if(changed) {
                applyLeds(device, dotIntensity, floodIntensity, mask);
            }
        }

        pipeline.stop();
        pipeline.wait();
    } catch(const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }

    return 0;
}

File: examples/python/Misc/IrLeds/ir_leds.py
#!/usr/bin/env python3
"""Test dot/flood IR LEDs with mono camera preview."""

import cv2
import depthai as dai

DOT_STEP = 0.1
FLOOD_STEP = 0.1


def print_drivers(device):
    print("Detected IR drivers:")
    for name, bus, addr in device.getIrDrivers():
        print(f"  name={name}  bus={bus}  addr=0x{addr:02x}")


def apply_leds(device, dot, flood, mask):
    ok_dot = device.setIrLaserDotProjectorIntensity(dot, mask)
    ok_flood = device.setIrFloodLightIntensity(flood, mask)
    print(f"dot={dot:.1f}  flood={flood:.1f}  mask={mask}  dot_ok={ok_dot}  flood_ok={ok_flood}")


def main():
    device = dai.Device()
    print(f"Connected to: {device.getDeviceName()}")
    print_drivers(device)

    with dai.Pipeline(device) as pipeline:
        mono_left = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_B)
        # mono_right = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_C)

        left_out = mono_left.requestFullResolutionOutput(type=dai.ImgFrame.Type.NV12)
        # right_out = mono_right.requestFullResolutionOutput(type=dai.ImgFrame.Type.NV12)

        left_queue = left_out.createOutputQueue()
        # right_queue = right_out.createOutputQueue()

        dot_intensity = 0.0
        flood_intensity = 0.0
        mask = -1  # all drivers/sides

        pipeline.start()
        apply_leds(device, dot_intensity, flood_intensity, mask)

        print("Controls:")
        print("  W/S : increase/decrease dot projector intensity")
        print("  A/D : increase/decrease flood light intensity")
        print("  1   : apply only to LEFT (mask 0x1)")
        print("  2   : apply only to RIGHT (mask 0x2)")
        print("  3   : apply to ALL (mask -1)")
        print("  P   : re-print IR drivers")
        print("  Q   : quit")

        while pipeline.isRunning():
            left = left_queue.get()
            # right = right_queue.get()
            assert isinstance(left, dai.ImgFrame)
            # assert isinstance(right, dai.ImgFrame)

            cv2.imshow("left", left.getCvFrame())
            # cv2.imshow("right", right.getCvFrame())

            key = cv2.waitKey(1)
            changed = False

            if key == ord('q') or key == 27:
                break
            elif key == ord('w'):
                dot_intensity = min(dot_intensity + DOT_STEP, 1.0)
                changed = True
            elif key == ord('s'):
                dot_intensity = max(dot_intensity - DOT_STEP, 0.0)
                changed = True
            elif key == ord('a'):
                flood_intensity = min(flood_intensity + FLOOD_STEP, 1.0)
                changed = True
            elif key == ord('d'):
                flood_intensity = max(flood_intensity - FLOOD_STEP, 0.0)
                changed = True
            elif key == ord('1'):
                mask = 0x1
                changed = True
            elif key == ord('2'):
                mask = 0x2
                changed = True
            elif key == ord('3'):
                mask = -1
                changed = True
            elif key == ord('p'):
                print_drivers(device)

            if changed:
                apply_leds(device, dot_intensity, flood_intensity, mask)

        pipeline.stop()


if __name__ == "__main__":
    main()
